### PR TITLE
Filter out auth chain queries that don't exist

### PR DESCRIPTION
**Pull request by realtyem**
Short version:
> It is kind of a logic fix in that you're doing less work by disregarding the sentinel value of 0. This has two improvements:

> 1. Avoiding expanding the chains dictionary unnecessarily.
> 2. Avoiding passing more arguments through the SQL engine (and to the database) which we know will result in no matches.

<details><summary>Long version:</summary>

In `_get_auth_chain_ids_using_cover_index_txn()`:
https://github.com/matrix-org/synapse/blob/c14a7de6af0b8a3cbf2e17afca1cab339bc5912d/synapse/storage/databases/main/event_federation.py#L300-L304

Seems to have a small logic problem, where the `max()` can evaluate to  `0`. The `chains` dict is passed into the following SQL block:

https://github.com/matrix-org/synapse/blob/c14a7de6af0b8a3cbf2e17afca1cab339bc5912d/synapse/storage/databases/main/event_federation.py#L314-L325

where it inhabits the `max_seq` variable. Inside the `event_auth_chains` database table, that column can never be `0`.

Filter out that `0` before it gets to the SQL.

The temporary metric attached produces results that appear like this(over less than 24 hours):
auth_chain_condition_hit

The metric commit isn't particularly efficient and has no long term relevance, so it will not be included in the final render.

</details>

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog).
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

Signed-off-by: Jason Little <realtyem@gmail.com>

### CODE REVIEW

File: synapse/storage/databases/main/event_federation.py
Comment (DMRobertson) [inline synapse/storage/databases/main/event_federation.py:1]: I agree that chain IDs are never 0. They come from a DB-backed sequence:

https://github.com/matrix-org/synapse/blob/12ca87f5eac06450abaf024e5f4906147d5322e3/synapse/storage/databases/main/events_worker.py#L293-L300

https://github.com/matrix-org/synapse/blob/25f43faa70f7cc58493b636c2702ae63395779dc/synapse/storage/schema/main/delta/59/04_event_auth_chains.sql.postgres#L16

which per https://www.postgresql.org/docs/16/sql-createsequence.html#id-1.9.3.81.6 should start at 1 and increase by 1. 

Therefore querying for events in chain 0 will never return anything.

Comment (DMRobertson) [reply synapse/storage/databases/main/event_federation.py:1]: Oh, your point is that the sequence_number in the chains is never 0?

Comment (DMRobertson) [reply synapse/storage/databases/main/event_federation.py:1]: It looks like these are also seeded at 1 and grow by 1:

https://github.com/matrix-org/synapse/blob/e9069c9f919685606506f04527332e83fbfa44d9/synapse/storage/databases/main/events.py#L941-L950

File: synapse/storage/databases/main/event_federation.py [hunk 1]
```diff
@@ -298,9 +298,13 @@ def _get_auth_chain_ids_using_cover_index_txn(
...
 
         # Add the initial set of chains, excluding the sequence corresponding to
-        # initial event.
+        # initial event. Ultimately, the chains dict will be what is pulled from the
+        # database, there is a chance that a sequence number here will end up being a
+        # '0', which doesn't exist. Don't bother sending that to the database query.
         for chain_id, seq_no in event_chains.items():
-            chains[chain_id] = max(seq_no - 1, chains.get(chain_id, 0))
+            max_sequence_result = max(seq_no - 1, chains.get(chain_id, 0))
+            if max_sequence_result > 0:
```
Comment (clokep) [inline synapse/storage/databases/main/event_federation.py:306]: I wonder if comparing to != 0 would be clearer?

```suggestion
            if max_sequence_result != 0:
```

File: synapse/storage/databases/main/event_federation.py [hunk 2]
```diff
@@ -298,9 +298,13 @@ def _get_auth_chain_ids_using_cover_index_txn(
                     )
 
         # Add the initial set of chains, excluding the sequence corresponding to
-        # initial event.
+        # initial event. Ultimately, the chains dict will be what is pulled from the
+        # database, there is a chance that a sequence number here will end up being a
+        # '0', which doesn't exist. Don't bother sending that to the database query.
         for chain_id, seq_no in event_chains.items():
-            chains[chain_id] = max(seq_no - 1, chains.get(chain_id, 0))
+            max_sequence_result = max(seq_no - 1, chains.get(chain_id, 0))
```
Comment (DMRobertson) [inline synapse/storage/databases/main/event_federation.py:305]: Do we understand when max_sequence_result is 0?

I think I've convinced myself that sequence numbers are strictly positive, and chains is a map to sequence numbers. Therefore we would need to have `seq_no == 1 and chain_id not in chains`. Is there any particular meaning to this situation?

Comment (DMRobertson) [reply synapse/storage/databases/main/event_federation.py:305]: I guess this corresponds to a brand-new chain that isn't the target of some other chain. Which sounds like the kind of thing that only happens at the start of a room? (cc @erikjohnston: is any of this sane?)

File: synapse/storage/databases/main/event_federation.py [hunk 3]
```diff
@@ -298,9 +298,13 @@ def _get_auth_chain_ids_using_cover_index_txn(
...
+            if max_sequence_result > 0:
+                chains[chain_id] = max_sequence_result
```
Comment (DMRobertson) [inline synapse/storage/databases/main/event_federation.py:307]: Is it possible that `chains` can now be empty? If so, will the query below fall over if we pass in an empty `chains.items()` to execute_values?

**Review by DMRobertson — COMMENTED** [review]
I think this is probably fine, but I was careful to double-check this since event auth chain logic is important to keep watertight. I would like @erikjohnston to double-check though.

File: synapse/storage/databases/main/event_federation.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/event_federation.py:306]: This is basically handling the case where an initial event is the first thing in the chain (and nothing else references it). This is actually pretty common, as we start a new chain whenever we see a new type / state key pair.

File: synapse/storage/databases/main/event_federation.py [hunk 4]
```diff
@@ -299,9 +299,13 @@ def _get_auth_chain_ids_using_cover_index_txn(
                     )
 
         # Add the initial set of chains, excluding the sequence corresponding to
-        # initial event.
+        # initial event. Ultimately, the chains dict will be what is pulled from the
+        # database, there is a chance that a sequence number here will end up being a
+        # '0', which doesn't exist. Don't bother sending that to the database query.
         for chain_id, seq_no in event_chains.items():
-            chains[chain_id] = max(seq_no - 1, chains.get(chain_id, 0))
+            max_sequence_result = max(seq_no - 1, chains.get(chain_id, 0))
```
Comment (erikjohnston) [inline synapse/storage/databases/main/event_federation.py:306]: I'd think I'd replace this change with something like:

```python
        for chain_id, seq_no in event_chains.items():
            # Check if the initial event is the first item in the chain. If so, then
            # there is nothing new to add from this chain.
            if seq_no == 1:
                continue
                
            chains[chain_id] = max(seq_no - 1, chains.get(chain_id, 0))
```

Maybe?

Comment (erikjohnston) [reply synapse/storage/databases/main/event_federation.py:306]: Not sure if that is clearer?

Comment (erikjohnston) [reply synapse/storage/databases/main/event_federation.py]: Thanks! Yeah, I think that comment is now redundant

**Review by erikjohnston — APPROVED** [review]
:heart:

**Pull request merged**